Python/Product/Debugger/DebugAdapterProcess.cs

```diff
@@ -15,41 +15,112 @@
 // permissions and limitations under the License.
 
 using System;
+using System.Collections.Generic;
 using System.Diagnostics;
 using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using Microsoft.PythonTools.Debugger.DebugEngine;
+using Microsoft.PythonTools.Infrastructure;
 using Microsoft.VisualStudio.Debugger.Interop.VSCodeDebuggerHost;
+using Newtonsoft.Json.Linq;
 
 namespace Microsoft.PythonTools.Debugger {
     public sealed class DebugAdapterProcess : ITargetHostProcess {
-        private readonly Process _process;
-        private DebugAdapterProcess(ProcessStartInfo psi) {
+        private const int _debuggerConnectionTimeout = 5000; // 5000 ms
+        private const int _connectionCloseTimeout = 5000; // 5000 ms
+
+        private Process _process;
+        private readonly Guid _processGuid;
+        private PythonDebugOptions _debugOptions;
+        private string _interpreterOptions;
+        private int _listenerPort = -1;
+        private Stream _stream;
+        private Socket _socket;
+        private ManualResetEvent _connectedEvent = new ManualResetEvent(false);
+
+        public DebugAdapterProcess() {
+            _processGuid = Guid.NewGuid();
+        }
+
+        public static DebugAdapterProcess Start(string launchJson) {
+            var debugProcess = new DebugAdapterProcess();
+            debugProcess.StartProcess(launchJson);
+            return debugProcess;
+        }
+
+        private void StartProcess(string launchJson) {
+            InitializeListenerSocket();
+
+            var json = JObject.Parse(launchJson);
+            var exe = json["exe"].Value<string>();
+            var args = json["args"].Value<string>();
+            var cwd = json["cwd"].Value<string>();
+            ParseOptions(json["options"].Value<string>());
+
+            List<string> argsList = new List<string> {
+                string.IsNullOrWhiteSpace(_interpreterOptions) ? "" : _interpreterOptions,
+                PythonToolsInstallPath.GetFile("ptvsd_launcher.py").AddQuotes(),
+                cwd.Trim('\\').AddQuotes(),
```

Comment:
This code should use `ProcessOutput` if possible. If not, then we need to take the code for argument quoting out of it, and make that usable on its own. The quoting is more involved than what's currently done here (consider what happens if the name of the current directory itself has a quote...). 

I'm fine merging this as is for now, just so that we get the working bits, and then clean it up in a separate commit. But then it shouldn't be defining an extension method, since that'll be going away anyway.